shared/common-adapters/plain-input.js.flow

```diff
@@ -0,0 +1,89 @@
+// @flow
+import * as React from 'react'
+import {type StylesCrossPlatform} from '../styles'
+import type {TextType} from './text'
+
+export type KeyboardType =
+  | 'default'
+  | 'email-address'
+  | 'numeric'
+  | 'phone-pad'
+  // iOS only
+  | 'ascii-capable'
+  | 'numbers-and-punctuation'
+  | 'url'
+  | 'number-pad'
+  | 'name-phone-pad'
+  | 'decimal-pad'
+  | 'twitter'
+  | 'web-search'
+  // Android Only
+  | 'visible-password'
+
+export type Props = {
+  autoFocus?: boolean,
+  className?: string,
+  disabled?: boolean,
+  // Resize in a flexbox-like fashion
+  flexable?: boolean,
+  maxLength?: number,
+  multiline?: boolean,
+  onBlur?: () => void,
+  onChangeText?: (text: string) => void,
+  onFocus?: () => void,
+  placeholder?: string,
+  rowsMin?: number,
+  rowsMax?: number,
+  style?: StylesCrossPlatform,
+  textType?: TextType,
+  type?: 'password' | 'text' | 'number',
+
+  // NOTE: Setting this prop will make the input controlled, which
+  // can impair performance badly on some platforms
+  value?: ?string,
+
+  /* Platform discrepancies */
+  // Maps to onSubmitEditing on native
+  onEnterKeyDown?: () => void,
+
+  // Desktop only
+  onClick?: (event: Event) => void,
+  onKeyDown?: (event: SyntheticKeyboardEvent<>, isComposingIME: boolean) => void,
+  onKeyUp?: (event: SyntheticKeyboardEvent<>, isComposingIME: boolean) => void,
+
+  // Mobile only
+  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters',
+  autoCorrect?: boolean,
+  keyboardType?: KeyboardType,
+  returnKeyType?: 'done' | 'go' | 'next' | 'search' | 'send',
+  selectTextOnFocus?: boolean,
+  onEndEditing?: () => void,
+}
+
+// Use this to mix your props with input props like type Props = PropsWithInput<{foo: number}>
+export type PropsWithInput<P> = {|
+  ...$Exact<Props>,
+  ...$Exact<P>,
+|}
+
+/**
+ * Flow does the work of making the default props nullable when instantiating
+ * this component, but doesn't go as far as letting the props be
+ * actually nullable in the type def. This complicates things when trying
+ * to make this compatible with PropsWithInput. So here we split up the
+ * internal type of Props from the public API, and 'lie' in this file
+ * by claiming that this component takes `Props` when the implementations
+ * use `InternalProps`.
+ * See more discussion here: https://github.com/facebook/flow/issues/1660
+ */
+export type DefaultProps = {
+  keyboardType: KeyboardType,
+  textType: TextType,
+}
+export type InternalProps = DefaultProps & Props
+declare export default class PlainInput extends React.Component<Props> {
+  static defaultProps: DefaultProps;
+  blur: () => void;
+  focus: () => void;
+  setValue: string => void;
```

Comment:
Hm, but we're getting rid of `transformText`, how would implementations set e.g. a default value without `setValue`?

Reply:
Hm OK, a `defaultValue` props sounds ok to me. I still think we don't want to burn the `setValue` bridge because it's an appropriate piece of API for this component to have and there's no alternative to it for things that just need to one-off set the value of the input.

Reply:
But `transformText` is intended to replace `setValue` -- we always want to set the selection while setting the text, as otherwise the behavior is inconsistent between iOS and Android. (This was the bug on chat input where selecting a @ completion in the middle of a line would move the selection to not the end of the completion.)

I'm afraid of leaving `setValue` (or `transformText`) in, as that might encourage clients to use it even when strictly not necessary, i.e. I want there to be a discussion the next time we think we need that functionality.

Reply:
Sounds good to me!